### DISCUSSION TITLE
Why does container.updateOverrides not work with a listener on a provider that is being watched?

### DISCUSSION THREAD

**Discussion by egonm12** [rrousselGit/riverpod#2465] [Q&A] [answered]
I would like to mock a watched provider in another provider so my test doesn't depend on the implementation details of the `Counter` in this example. However, updating the overridden provider doesn't seem to behave as expected with a listener. When I remove the listener, and work with `container.read` and `container.pump` it seems to work as expected. Is this something that is simply not possible or am I doing it wrong?

Consider these providers:
```
@riverpod
bool isCounterOdd(IsCounterOddRef ref) {
  final counter = ref.watch(counterProvider);

  return counter.isOdd;
}

@riverpod
class Counter extends _$Counter {
  @override
  int build() => 0;

  void increment() {
    state = state + 1;
  }
}
```

With the following mock:
```
class MockAutoDisposeNotifier<T> extends AutoDisposeNotifier<T> with Mock {
  MockAutoDisposeNotifier(this._build) : super();

  final T Function() _build;

  @override
  T build() => _build();
}

class CounterMock extends MockAutoDisposeNotifier<int> implements Counter {
  CounterMock(super._build);
}
```

The following test fails, because the second call is never made.
```
test('some test description', () {
  final listener = Listener<bool>();
  final container = ProviderContainer(
    overrides: [
      counterProvider.overrideWith(() => CounterMock(() => 0)),
    ],
  )..listen(
      isCounterOddProvider,
      listener.call,
      fireImmediately: true,
    );

  verify(() => listener(null, false)).called(1);
  verifyNoMoreInteractions(listener);

  container.updateOverrides([
    counterProvider.overrideWith(() => CounterMock(() => 1)),
  ]);

  container.refresh(isCounterOddProvider);
  verify(() => listener(false, true)).called(1);
});
```
This works fine however:
```
test('some test description', () async {
  final container = ProviderContainer(
    overrides: [
      counterProvider.overrideWith(() => CounterMock(() => 0)),
    ],
  );

  expect(container.read(isCounterOddProvider), false);

  await container.pump();

  container.updateOverrides([
    counterProvider.overrideWith(() => CounterMock(() => 1)),
  ]);
  expect(container.read(isCounterOddProvider), true);
});
```

**Comment by rrousselGit** [comment, marked as answer]
overrideWith isn't reapplied until a provider is fully destroyed.  
A different behavior would make using overrides inside widgets very problematic.

Say your widget creates a `ProviderScope` with overrides rebuilds: That would destroy the state of all the overridden providers.

That's obviously undesired, so Riverpod keeps the previous behavior instead.

**Reply by egonm12** [reply]
Ok, your explanation makes sense. However, I can't make the connection between your explanation and the behaviour in the test. E.g. this test also fails, but when I remove the `.listen(..)` it works fine again. 🤔 
```
test('some test description', () async {
  final listener = Listener<bool>();
  final container = ProviderContainer(
    overrides: [
      counterProvider.overrideWith(() => CounterMock(() => 0)),
    ],
  )..listen( // removing this will let this test pass again.
      isCounterOddProvider,
      listener.call,
      fireImmediately: true,
    );

  expect(container.read(isCounterOddProvider), false);

  await container.pump();

  container.updateOverrides([
    counterProvider.overrideWith(() => CounterMock(() => 1)),
  ]);
  expect(container.read(isCounterOddProvider), true);
});
```

**Reply by rrousselGit** [reply]
It's about autoDispose.
If the provider got disposed, then the new behavior takes over.

**Reply by egonm12** [reply]
How does adding a listener to a provider have an effect on the disposal of the provider? I don't get why adding `.listen(...)` to the `container` doesn't affect the watched provider's value when updating the override of that provider, while it does work fine without the `.listen(...)`.

**Reply by rrousselGit** [reply]
AutoDispose applies when a provider isn't listened

**Reply by egonm12** [reply]
Clear, thank you!
